File: src/config/store.ts
/**
 * 配置存取层：读写 ~/.cc-mini/providers.json，生成默认配置，构造 env map。
 *
 * 铁律：这里写的是工具自己的数据文件（providers.json），**绝不**碰 ~/.claude/*。
 * 格式与现版 PowerShell 完全兼容；写入 UTF-8 无 BOM、2 空格缩进。
 */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

import { KNOWN_KEYS, type Lang, type Provider, type Store } from './types.js';

export { KNOWN_KEYS } from './types.js';
export type { Lang, ManagedKey, Provider, Store } from './types.js';

export interface StorePaths {
  dir: string;
  file: string;
}

/** 解析存储路径。`storeDir` 来自 --store-dir（测试用），默认 ~/.cc-mini。 */
export function resolveStorePaths(storeDir?: string): StorePaths {
  const dir = storeDir && storeDir.trim() ? storeDir : join(homedir(), '.cc-mini');
  return { dir, file: join(dir, 'providers.json') };
}

const nonEmpty = (v: unknown): v is string => typeof v === 'string' && v.trim() !== '';

/**
 * 默认配置：官方 + DeepSeek + 智谱GLM + 小米MiMo（密钥空）。
 * 官方档带 `builtin: 'official'`（评审①），其它供应商是专有名词、不翻译。
 */
export function defaultStore(): Store {
  return {
    current: '官方',
    lang: 'zh',
    providers: [
      { name: '官方', note: '', builtin: 'official', env: {} },
      {
        name: 'DeepSeek',
        note: '',
        env: {
          ANTHROPIC_BASE_URL: 'https://api.deepseek.com/anthropic',
          ANTHROPIC_DEFAULT_OPUS_MODEL: 'deepseek-v4-pro',
          ANTHROPIC_DEFAULT_SONNET_MODEL: 'deepseek-v4-pro',
          ANTHROPIC_DEFAULT_HAIKU_MODEL: 'deepseek-v4-flash',
          CLAUDE_CODE_EFFORT_LEVEL: 'max',
        },
      },
      {
        name: '智谱GLM',
        note: '',
        env: {
          ANTHROPIC_BASE_URL: 'https://open.bigmodel.cn/api/anthropic',
          ANTHROPIC_DEFAULT_OPUS_MODEL: 'GLM-4.7',
          ANTHROPIC_DEFAULT_SONNET_MODEL: 'GLM-4.7',
          ANTHROPIC_DEFAULT_HAIKU_MODEL: 'glm-4.5-air',
        },
      },
      {
        name: '小米MiMo',
        note: '',
        env: {
          ANTHROPIC_BASE_URL: 'https://api.xiaomimimo.com/anthropic',
          ANTHROPIC_DEFAULT_OPUS_MODEL: 'mimo-v2.5-pro',
          ANTHROPIC_DEFAULT_SONNET_MODEL: 'mimo-v2.5-pro',
          ANTHROPIC_DEFAULT_HAIKU_MODEL: 'mimo-v2.5-pro',
        },
      },
    ],
  };
}

/**
 * 把任意解析结果规整为合法 Store。
 *
 * 字段级**宽松容错**（缺 lang / 缺 note / 缺 builtin / 缺 env 都不报错），保持既有数据格式兼容；
 * 但结构级**严格校验**：顶层必须是对象、`providers` 必须是数组、每个配置的 name/env 结构必须合法
 * —— 否则抛 StoreError('format')。
 * 这是为了堵住「语法合法但结构损坏的 JSON 被静默规整成空 providers、用户一保存就覆盖丢数据」的坑（[P1]）。
 */
function normalizeStore(raw: unknown, file: string): Store {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) throw new StoreError('format', file);
  const obj = raw as Record<string, unknown>;
  if (!Array.isArray(obj.providers)) throw new StoreError('format', file);
  const providers: Provider[] = obj.providers.map((p) => normalizeProvider(p, file));
  const lang: Lang | undefined = obj.lang === 'en' ? 'en' : obj.lang === 'zh' ? 'zh' : undefined;
  const current = typeof obj.current === 'string' ? obj.current : (providers[0]?.name ?? '');
  // 只认已知模式，未知值视为关闭（不回写）。字段顺序须与 Go 版一致：…providers, update?
  const update = obj.update === 'notify' || obj.update === 'auto' ? obj.update : undefined;
  return { current, ...(lang ? { lang } : {}), providers, ...(update ? { update } : {}) };
}

function normalizeProvider(raw: unknown, file: string): Provider {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) throw new StoreError('format', file);
  const p = raw as Record<string, unknown>;
  if (typeof p.name !== 'string') throw new StoreError('format', file);
  if (p.note !== undefined && typeof p.note !== 'string') throw new StoreError('format', file);
  if (p.builtin !== undefined && typeof p.builtin !== 'string') throw new StoreError('format', file);
  if (p.env !== undefined && (!p.env || typeof p.env !== 'object' || Array.isArray(p.env))) {
    throw new StoreError('format', file);
  }
  const name = p.name;
  const note = typeof p.note === 'string' ? p.note : '';
  const builtin = typeof p.builtin === 'string' ? p.builtin : undefined;
  const env: Record<string, string> = {};
  if (p.env) {
    for (const [k, v] of Object.entries(p.env as Record<string, unknown>)) {
      if (typeof v !== 'string') throw new StoreError('format', file);
      env[k] = v;
    }
  }
  return { name, note, ...(builtin ? { builtin } : {}), env };
}

/**
 * 配置文件存在但不可用时抛出。调用方据此给出友好提示并退出，
 * **绝不**静默重建/覆盖——那会清掉用户的明文密钥（违背「不碰用户数据」初心）。
 *   - `read`  ：读文件失败（权限/磁盘/同名目录 EISDIR …）[P2]
 *   - `parse` ：JSON 语法损坏 [上一轮已修]
 *   - `format`：JSON 语法合法但结构损坏（顶层非对象 / providers 非数组）[P1]
 */
export type StoreErrorKind = 'read' | 'parse' | 'format';

export class StoreError extends Error {
  constructor(
    public readonly kind: StoreErrorKind,
    public readonly file: string,
  ) {
    super(`store ${kind} error: ${file}`);
    this.name = 'StoreError';
  }
}

/** 读配置；文件不存在则生成默认并落盘后返回；文件不可用则抛 StoreError（绝不覆盖）。 */
export function loadStore(paths: StorePaths): Store {
  let text: string;
  try {
    text = readFileSync(paths.file, 'utf-8'); // [P2] 权限/磁盘/EISDIR 等
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code !== 'ENOENT') throw new StoreError('read', paths.file);
    const store = defaultStore();
    saveStore(paths, store);
    return store;
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    throw new StoreError('parse', paths.file);
  }
  return normalizeStore(parsed, paths.file); // [P1] 结构校验在内部，失败抛 'format'
}

/**
 * 只读探测 lang，**不生成文件**（用于 --help/--version 在 parse 前定语言，避免副作用）。
 * 文件不存在/解析失败都返回 undefined。
 */
export function peekStoreLang(paths: StorePaths): Lang | undefined {
  try {
    if (!existsSync(paths.file)) return undefined;
    const raw = JSON.parse(readFileSync(paths.file, 'utf-8')) as { lang?: unknown };
    return raw.lang === 'en' ? 'en' : raw.lang === 'zh' ? 'zh' : undefined;
  } catch {
    return undefined;
  }
}

/** 写配置：UTF-8 无 BOM、2 空格缩进（Node 默认 utf-8 不带 BOM，与现版一致）。 */
export function saveStore(paths: StorePaths, store: Store): void {
  if (!existsSync(paths.dir)) mkdirSync(paths.dir, { recursive: true });
  writeFileSync(paths.file, `${JSON.stringify(store, null, 2)}\n`, 'utf-8');
}

/**
 * 是否官方档。优先认稳定标识 `builtin === 'official'`（评审①）；
 * 老文件没有 builtin 时，仅将「中文名为官方 + env 为空」视为官方档。
 * 这样旧数据仍兼容，而手动填了第三方地址/密钥的同名配置不会被误判为登录态。
 */
export function isOfficial(p: Provider): boolean {
  if (p.builtin) return p.builtin === 'official';
  return p.name === '官方' && Object.keys(p.env).length === 0;
}

/**
 * [P1] 编辑保存后修正身份：官方档（builtin='official' = 登录态、空 env）一旦被配成真实第三方
 * （env 非空，有了 base/key 等），就清掉 builtin —— 否则会继续被当登录态、跳过缺密钥警告。
 */
export function reconcileBuiltin(p: Provider): void {
  if (p.builtin === 'official' && Object.keys(p.env).length > 0) {
    delete p.builtin;
  }
}

/** 删除配置等操作后修正默认指向：优先剩余官方档，其次第一项；没有配置则置空。 */
export function reconcileCurrent(store: Store): void {
  if (store.providers.some((p) => p.name === store.current)) return;
  store.current = (store.providers.find(isOfficial) ?? store.providers[0])?.name ?? '';
}

/** 取配置的 env map（即 provider.env，保证非空对象）。 */
export function getProviderEnvMap(p: Provider): Record<string, string> {
  return p.env ?? {};
}

/**
 * 由一组字段构造 provider.env：按 KNOWN_KEYS 顺序、丢弃空白值。
 * 对齐现版 Build-ProviderEnv。`[1m]` 等后缀属于自由文本，原样保留（见 plan §3.1.1）。
 */
export function buildProviderEnv(fields: Record<string, string | undefined>): Record<string, string> {
  const env: Record<string, string> = {};
  for (const key of KNOWN_KEYS) {
    const v = fields[key];
    if (nonEmpty(v)) env[key] = v.trim();
  }
  return env;
}

/** 配置的密钥状态（语义枚举，不含界面文案；翻译交给 i18n 层，评审①）。 */
export type KeyState = 'official' | 'noKey' | 'apiKey' | 'hasToken';

export interface ProviderState {
  key: KeyState;
  effort?: string;
}

export function getProviderState(p: Provider): ProviderState {
  const map = getProviderEnvMap(p);
  const effort = nonEmpty(map.CLAUDE_CODE_EFFORT_LEVEL) ? map.CLAUDE_CODE_EFFORT_LEVEL : undefined;
  if (isOfficial(p)) return { key: 'official', ...(effort ? { effort } : {}) };
  const hasTok = nonEmpty(map.ANTHROPIC_AUTH_TOKEN);
  const hasKey = nonEmpty(map.ANTHROPIC_API_KEY);
  const key: KeyState = !hasTok && !hasKey ? 'noKey' : hasKey ? 'apiKey' : 'hasToken';
  return { key, ...(effort ? { effort } : {}) };
}

/** 按 name 找配置。 */
export function findProvider(store: Store, name: string): Provider | undefined {
  return store.providers.find((p) => p.name === name);
}

/**
 * 名称去重：同名被【其它】配置占用时追加 “ 2/3/…”。`exclude` 是正在编辑的本条（排除自身）。
 * 对齐现版 Resolve-UniqueName。
 */
export function resolveUniqueName(store: Store, name: string, exclude?: Provider): string {
  const existing = store.providers.filter((p) => p !== exclude).map((p) => p.name);
  if (!existing.includes(name)) return name;
  let i = 2;
  while (existing.includes(`${name} ${i}`)) i += 1;
  return `${name} ${i}`;
}

/** 语言：providers.json 的 lang 字段，缺省视为 zh。 */
export function getLang(store: Store): Lang {
  return store.lang === 'en' ? 'en' : 'zh';
}

export function setLang(store: Store, lang: Lang): void {
  store.lang = lang;
}

File: internal/config/store.go
package config

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// StorePaths 是 store 目录与文件的解析结果。
type StorePaths struct {
	Dir  string
	File string
}

// ResolveStorePaths 解析存储路径。storeDir 来自 --store-dir（测试用），默认 ~/.cc-mini。
func ResolveStorePaths(storeDir string) StorePaths {
	dir := storeDir
	if strings.TrimSpace(dir) == "" {
		home, _ := os.UserHomeDir()
		dir = filepath.Join(home, ".cc-mini")
	}
	return StorePaths{Dir: dir, File: filepath.Join(dir, "providers.json")}
}

func nonEmpty(v string) bool { return strings.TrimSpace(v) != "" }

// DefaultStore 返回默认配置：官方 + DeepSeek + 智谱GLM + 小米MiMo（密钥空）。
// 官方档带 builtin="official"（评审①），其它供应商是专有名词、不翻译。
func DefaultStore() *Store {
	return &Store{
		Current: "官方",
		Lang:    LangZH,
		Providers: []Provider{
			{Name: "官方", Note: "", Builtin: "official", Env: map[string]string{}},
			{Name: "DeepSeek", Note: "", Env: map[string]string{
				"ANTHROPIC_BASE_URL":             "https://api.deepseek.com/anthropic",
				"ANTHROPIC_DEFAULT_OPUS_MODEL":   "deepseek-v4-pro",
				"ANTHROPIC_DEFAULT_SONNET_MODEL": "deepseek-v4-pro",
				"ANTHROPIC_DEFAULT_HAIKU_MODEL":  "deepseek-v4-flash",
				"CLAUDE_CODE_EFFORT_LEVEL":       "max",
			}},
			{Name: "智谱GLM", Note: "", Env: map[string]string{
				"ANTHROPIC_BASE_URL":             "https://open.bigmodel.cn/api/anthropic",
				"ANTHROPIC_DEFAULT_OPUS_MODEL":   "GLM-4.7",
				"ANTHROPIC_DEFAULT_SONNET_MODEL": "GLM-4.7",
				"ANTHROPIC_DEFAULT_HAIKU_MODEL":  "glm-4.5-air",
			}},
			{Name: "小米MiMo", Note: "", Env: map[string]string{
				"ANTHROPIC_BASE_URL":             "https://api.xiaomimimo.com/anthropic",
				"ANTHROPIC_DEFAULT_OPUS_MODEL":   "mimo-v2.5-pro",
				"ANTHROPIC_DEFAULT_SONNET_MODEL": "mimo-v2.5-pro",
				"ANTHROPIC_DEFAULT_HAIKU_MODEL":  "mimo-v2.5-pro",
			}},
		},
	}
}

// StoreErrorKind 区分配置文件不可用的原因。
type StoreErrorKind string

const (
	// ErrRead 读文件失败（权限/磁盘/同名目录等）。
	ErrRead StoreErrorKind = "read"
	// ErrParse JSON 语法损坏。
	ErrParse StoreErrorKind = "parse"
	// ErrFormat JSON 语法合法但结构损坏（顶层非对象 / providers 非数组等）。
	ErrFormat StoreErrorKind = "format"
)

// StoreError 在配置文件存在但不可用时返回。调用方据此友好提示并退出，
// 绝不静默重建/覆盖——那会清掉用户的明文密钥（违背「不碰用户数据」初心）。
type StoreError struct {
	Kind StoreErrorKind
	File string
}

func (e *StoreError) Error() string {
	return fmt.Sprintf("store %s error: %s", e.Kind, e.File)
}

// Load 读配置；文件不存在则生成默认并落盘后返回；文件不可用则返回 *StoreError（绝不覆盖）。
func Load(p StorePaths) (*Store, error) {
	data, err := os.ReadFile(p.File)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			s := DefaultStore()
			if werr := Save(p, s); werr != nil {
				return nil, werr
			}
			return s, nil
		}
		return nil, &StoreError{Kind: ErrRead, File: p.File}
	}
	var raw any
	if jerr := json.Unmarshal(data, &raw); jerr != nil {
		return nil, &StoreError{Kind: ErrParse, File: p.File}
	}
	return normalizeStore(raw, p.File)
}

// normalizeStore 把任意解析结果规整为合法 Store。
//
// 字段级宽松容错（缺 lang/note/builtin/env 都不报错）保持兼容；
// 结构级严格校验（顶层须对象、providers 须数组、每条 name/env 结构须合法）否则返回 ErrFormat。
// 堵住「语法合法但结构损坏的 JSON 被静默规整成空 providers、用户一保存就覆盖丢数据」的坑。
func normalizeStore(raw any, file string) (*Store, error) {
	obj, ok := raw.(map[string]any)
	if !ok {
		return nil, &StoreError{Kind: ErrFormat, File: file}
	}
	provsRaw, ok := obj["providers"].([]any)
	if !ok {
		return nil, &StoreError{Kind: ErrFormat, File: file}
	}
	providers := make([]Provider, 0, len(provsRaw))
	for _, pr := range provsRaw {
		p, err := normalizeProvider(pr, file)
		if err != nil {
			return nil, err
		}
		providers = append(providers, p)
	}
	var lang Lang
	if l, ok := obj["lang"].(string); ok {
		switch l {
		case "en":
			lang = LangEN
		case "zh":
			lang = LangZH
		}
	}
	current := ""
	if c, ok := obj["current"].(string); ok {
		current = c
	} else if len(providers) > 0 {
		current = providers[0].Name
	}
	return &Store{Current: current, Lang: lang, Providers: providers}, nil
}

func normalizeProvider(raw any, file string) (Provider, error) {
	formatErr := func() (Provider, error) { return Provider{}, &StoreError{Kind: ErrFormat, File: file} }
	p, ok := raw.(map[string]any)
	if !ok {
		return formatErr()
	}
	name, ok := p["name"].(string)
	if !ok {
		return formatErr()
	}
	note := ""
	if nv, exists := p["note"]; exists {
		s, ok := nv.(string)
		if !ok {
			return formatErr()
		}
		note = s
	}
	builtin := ""
	if bv, exists := p["builtin"]; exists {
		s, ok := bv.(string)
		if !ok {
			return formatErr()
		}
		builtin = s
	}
	env := map[string]string{}
	if ev, exists := p["env"]; exists {
		m, ok := ev.(map[string]any)
		if !ok {
			return formatErr()
		}
		for k, vv := range m {
			s, ok := vv.(string)
			if !ok {
				return formatErr()
			}
			env[k] = s
		}
	}
	return Provider{Name: name, Note: note, Builtin: builtin, Env: env}, nil
}

// PeekStoreLang 只读探测 lang，不生成文件（用于 --help/--version 在 parse 前定语言，避免副作用）。
// 文件不存在/解析失败都返回空。
func PeekStoreLang(p StorePaths) Lang {
	data, err := os.ReadFile(p.File)
	if err != nil {
		return ""
	}
	var raw struct {
		Lang string `json:"lang"`
	}
	if json.Unmarshal(data, &raw) != nil {
		return ""
	}
	switch raw.Lang {
	case "en":
		return LangEN
	case "zh":
		return LangZH
	}
	return ""
}

// Save 写配置：UTF-8 无 BOM、2 空格缩进、不转义 HTML、尾随换行（与 npm 版一致）。
func Save(p StorePaths, store *Store) error {
	if err := os.MkdirAll(p.Dir, 0o755); err != nil {
		return &StoreError{Kind: ErrRead, File: p.File}
	}
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	enc.SetIndent("", "  ")
	if err := enc.Encode(store); err != nil {
		return err
	}
	// enc.Encode 已追加一个换行，与 npm 版 `JSON.stringify(...)+"\n"` 一致。
	if err := os.WriteFile(p.File, buf.Bytes(), 0o644); err != nil {
		return &StoreError{Kind: ErrRead, File: p.File}
	}
	return nil
}

// IsOfficial 判断是否官方档。优先认稳定标识 builtin=="official"（评审①）；
// 老文件没有 builtin 时，仅将「中文名为官方 + env 为空」视为官方档。
func IsOfficial(p Provider) bool {
	if p.Builtin != "" {
		return p.Builtin == "official"
	}
	return p.Name == "官方" && len(p.Env) == 0
}

// ReconcileBuiltin 编辑保存后修正身份：官方档（builtin=official、空 env）一旦被配成真实第三方
// （env 非空），就清掉 builtin —— 否则会继续被当登录态、跳过缺密钥警告。
func ReconcileBuiltin(p *Provider) {
	if p.Builtin == "official" && len(p.Env) > 0 {
		p.Builtin = ""
	}
}

// ReconcileCurrent 删除等操作后修正默认指向：优先剩余官方档，其次第一项；没有配置则置空。
func ReconcileCurrent(store *Store) {
	for _, p := range store.Providers {
		if p.Name == store.Current {
			return
		}
	}
	for _, p := range store.Providers {
		if IsOfficial(p) {
			store.Current = p.Name
			return
		}
	}
	if len(store.Providers) > 0 {
		store.Current = store.Providers[0].Name
	} else {
		store.Current = ""
	}
}

// GetProviderEnvMap 取配置的 env map（保证非 nil）。
func GetProviderEnvMap(p Provider) map[string]string {
	if p.Env == nil {
		return map[string]string{}
	}
	return p.Env
}

// BuildProviderEnv 由一组字段构造 provider.env：按 KnownKeys 顺序、丢弃空白值（trim 后存）。
// `[1m]` 等后缀属于自由文本，原样保留（见 plan §3.1.1）。
func BuildProviderEnv(fields map[string]string) map[string]string {
	env := map[string]string{}
	for _, key := range knownKeys {
		v := fields[key]
		if nonEmpty(v) {
			env[key] = strings.TrimSpace(v)
		}
	}
	return env
}

// KeyState 是配置的密钥状态（语义枚举，不含界面文案；翻译交给 i18n 层，评审①）。
type KeyState string

const (
	KeyOfficial KeyState = "official"
	KeyNone     KeyState = "noKey"
	KeyAPIKey   KeyState = "apiKey"
	KeyToken    KeyState = "hasToken"
)

// ProviderState 是配置的运行状态：密钥种类 + effort。
type ProviderState struct {
	Key    KeyState
	Effort string
}

// GetProviderState 计算配置状态，对齐 npm 版 getProviderState。
func GetProviderState(p Provider) ProviderState {
	m := GetProviderEnvMap(p)
	effort := ""
	if nonEmpty(m["CLAUDE_CODE_EFFORT_LEVEL"]) {
		effort = m["CLAUDE_CODE_EFFORT_LEVEL"]
	}
	if IsOfficial(p) {
		return ProviderState{Key: KeyOfficial, Effort: effort}
	}
	hasTok := nonEmpty(m["ANTHROPIC_AUTH_TOKEN"])
	hasKey := nonEmpty(m["ANTHROPIC_API_KEY"])
	key := KeyToken
	switch {
	case !hasTok && !hasKey:
		key = KeyNone
	case hasKey:
		key = KeyAPIKey
	}
	return ProviderState{Key: key, Effort: effort}
}

// FindProvider 按 name 找配置；找不到返回 nil。
func FindProvider(store *Store, name string) *Provider {
	for i := range store.Providers {
		if store.Providers[i].Name == name {
			return &store.Providers[i]
		}
	}
	return nil
}

// ResolveUniqueName 名称去重：同名被【其它】配置占用时追加 " 2/3/…"。
// exclude 是正在编辑的本条（按指针排除自身，传 nil 表示不排除）。对齐 npm 版 resolveUniqueName。
func ResolveUniqueName(store *Store, name string, exclude *Provider) string {
	existing := make(map[string]bool)
	for i := range store.Providers {
		if exclude != nil && &store.Providers[i] == exclude {
			continue
		}
		existing[store.Providers[i].Name] = true
	}
	if !existing[name] {
		return name
	}
	for i := 2; ; i++ {
		cand := fmt.Sprintf("%s %d", name, i)
		if !existing[cand] {
			return cand
		}
	}
}

// GetLang 取语言：store.lang 字段，缺省视为 zh。
func GetLang(store *Store) Lang {
	if store.Lang == LangEN {
		return LangEN
	}
	return LangZH
}

// SetLang 设置语言。
func SetLang(store *Store, lang Lang) {
	store.Lang = lang
}
